### PR TITLE
New feature for permissions/access control using ItemAttribute

### PR DESCRIPTION
**Pull request by trannamtrung1st**
I need to add permissions/access control functionality to my project so I create this PR to update that new feature here. Please take a look if you have time.
I've tested for most situations and it still worked fine. Give me comments if there is any problem.

Many thanks.

### CODE REVIEW

File: elFinder.NetCore.Web/Controllers/FileSystemController.cs [hunk 1]
```diff
@@ -35,12 +36,32 @@ private Connector GetConnector()
...
                 Alias = "Files", // Beautiful name given to the root/home folder
                 //MaxUploadSizeInKb = 2048, // Limit imposed to user uploaded file <= 2048 KB
-                //LockedFolders = new List<string>(new string[] { "Folder1" })
+                //DefaultAttribute = new ItemAttribute // Restrict access to all items in root, use ItemAttributes to enable access for specific items
+                //{
+                //    Locked = true,
+                //    Read = false,
+                //    Write = false
+                //},
+                ItemAttributes = new HashSet<NamedItemAttribute>()
```
Comment (trannamtrung1st) [inline elFinder.NetCore.Web/Controllers/FileSystemController.cs:51]: Please see **elFinder.NetCore/ItemAttribute.cs**

File: elFinder.NetCore/Drivers/FileSystem/FileSystemDriver.cs [hunk 2]
```diff
@@ -302,7 +303,11 @@ public async Task<JsonResult> InitAsync(FullPath path, IEnumerable<string> mimeT
                     root = Roots.First();
                 }
 
-                path = new FullPath(root, new FileSystemDirectory(root.StartDirectory ?? root.RootDirectory), null);
+                if (Directory.Exists(root.StartDirectory))
+                    path = new FullPath(root, new FileSystemDirectory(root.StartDirectory), null);
+
+                if (path == null || path.Directory.GetReadFlag(root) == 0)
+                    path = new FullPath(root, new FileSystemDirectory(root.RootDirectory), null);
```
Comment (trannamtrung1st) [inline elFinder.NetCore/Drivers/FileSystem/FileSystemDriver.cs:310]: Fallback to root if StartDirectory does not exists or does not have read permission

File: elFinder.NetCore/Extensions/IDirectoryExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace elFinder.NetCore.Extensions
+{
+    public static class IDirectoryExtensions
+    {
+        public static byte GetLockedFlag(this IDirectory directory, RootVolume volume)
```
Comment (trannamtrung1st) [inline elFinder.NetCore/Extensions/IDirectoryExtensions.cs:12]: The idea for these Get...Flag methods is to use the below priority (high to low) to determine the flag value:
- Check RootVolume settings (IsReadOnly, IsLocked)
- Check if there is any specific NamedItemAttribute for current File/Directory 
- Check if there is any NamedItemAttribute for parents (recursive until root is met) of current File/Directory.
- Use the default ItemAttribute set in RootVolume

**Comment by trannamtrung1st** [comment]
Hi @gordon-matt , @fsmirne,
You all may want to take a look at this. I'll be very happy to see this works officially.

Thanks.

File: elFinder.NetCore.Web/Startup.cs [hunk 4]
```diff
@@ -19,7 +20,7 @@ public static string MapPath(string path, string basePath = null)
             }
 
             path = path.Replace("~/", "").TrimStart('/').Replace('/', '\\');
-            return Path.Combine(basePath, path);
+            return PathHelper.GetFullPathNormalized(Path.Combine(basePath, path));
```
Comment (trannamtrung1st) [inline elFinder.NetCore.Web/Startup.cs:23]: See **elFinder.NetCore/Helpers/PathHelper.cs**.
Use this method to ensure all paths are normalized.
**Current**: "path/a/b" is different from "path\a/b", "path/a" is different from "path/a/"

File: elFinder.NetCore/Drivers/FileSystem/FileSystemFile.cs [hunk 5]
```diff
@@ -8,22 +8,17 @@ namespace elFinder.NetCore.Drivers.FileSystem
 {
     public class FileSystemFile : IFile
     {
-        private readonly string filePath;
+        private FileInfo fileInfo;
```
Comment (trannamtrung1st) [inline elFinder.NetCore/Drivers/FileSystem/FileSystemFile.cs:11]: - Use FileInfo object to reuse and ensure an operation (unit of work) uses the same states of file.
- Gain performance by bypass File... checking. See: [FileInfo Remarks](https://docs.microsoft.com/en-us/dotnet/api/system.io.fileinfo?view=net-5.0#remarks)

File: elFinder.NetCore/Drivers/FileSystem/FileSystemFile.cs [hunk 6]
```diff
@@ -110,5 +114,11 @@ private static void EnsureGarbageCollectorCalled()
             GC.Collect();
             GC.WaitForPendingFinalizers();
         }
+
+        public Task RefreshAsync()
```
Comment (trannamtrung1st) [inline elFinder.NetCore/Drivers/FileSystem/FileSystemFile.cs:118]: - Explicitly refresh the state of current FileInfo if necessary.

File: elFinder.NetCore/Drivers/FileSystem/FileSystemDirectory.cs [hunk 7]
```diff
@@ -97,5 +105,11 @@ public void MoveTo(string destDirName)
         {
             directoryInfo.MoveTo(destDirName);
         }
+
+        public Task RefreshAsync()
+        {
```
Comment (trannamtrung1st) [inline elFinder.NetCore/Drivers/FileSystem/FileSystemDirectory.cs:110]: - Explicitly refresh the state of current DirectoryInfo if necessary.

**Comment by gordon-matt** [comment]
Thanks for your contribution. That's a lot of changes for me to review. I will try to find time soon.

**Comment by trannamtrung1st** [comment]
> Thanks for your contribution. That's a lot of changes for me to review. I will try to find time soon.

Glad to here that 👍

**Comment by gordon-matt** [comment]
Thanks. I have now implemented your work.. I did it a little bit differently, so I will close this PR instead of merging it. You can get the latest code and tell me if everything is working as you expected. I've only done some basic testing, but it seems okay so far.

Good work. Cam on ban nhieu lam  ;-)

**Comment by trannamtrung1st** [comment]
> Thanks. I have now implemented your work.. I did it a little bit differently, so I will close this PR instead of merging it. You can get the latest code and tell me if everything is working as you expected. I've only done some basic testing, but it seems okay so far.
> 
> Good work. Cam on ban nhieu lam ;-)

You're welcome. Thanks for the work, I will update the latest code.

**Comment by gordon-matt** [comment]
I will try to find time soon to look at your other PRs before I release a new version.

@fsmirne and @brunomel, please review the changes I made to the file system driver and see if you need to make similar changes in your Azure drivers.

**Comment by gordon-matt** [comment]
New version released and it includes PR #62 as well

**Pull request closed without merging**